Sema: preserve field alignment in union pointer captures

schmee opened the pull request:
Fixes https://github.com/ziglang/zig/issues/18660.

mlugg commented on test/behavior/switch.zig:
```diff
@@ -574,6 +574,69 @@ test "switch prongs with cases with identical payload types" {
+        fn doTheTest() !void {
+            const u = U{ .a = 1 };
+            switch (u) {
+                .a => |*a| try expectEqual(*align(8) const u8, @TypeOf(a)),
```
When checking types are what was expected, use `comptime assert` instead so failures emit a compile error.

mlugg commented on test/behavior/switch.zig:
```diff
@@ -574,6 +574,69 @@ test "switch prongs with cases with identical payload types" {
+                .a, .b => |*p| try expectEqual(*align(4) const u8, @TypeOf(p)),
+                .c => |*p| {
+                    _ = p;
+                    @panic("fail");
```
These should return an error to make the test fail rather than panic. It's okay, and in fact encouraged, for a behavior test to verify other behaviours it's coincidentally using.

mlugg reviewed (changes requested):
Nice patch, some small nitpicks.

schmee commented:
@mlugg thanks for the review! I've addressed your comments, and did the changes you suggested to other tests in `switch.zig` as well to nudge any future readers in the right direction 👍

andrewrk commented:
Thanks @schmee and for the review @mlugg :+1:
